Purchases: Add `state.purchases.items` reducer

drewblaisdell opened the pull request:
Addresses part of https://github.com/Automattic/wp-calypso/issues/5046.

This PR copies the reducer and its tests from `lib/purchases/reducer` into `state/purchases/` and updates them such that:

- The tests test the reducer, instead of a Flux singleton.
- The reducer is exported in a function assigned as `items`.

It also sorts `state/action-types`, which was not entirely alphabetized.

This PR does not remove the old reducer or Flux store. Nor does it import the newly updated reducer into the combined global reducer. This will happen in a separate PR to keep the PR size manageable.

**Testing**
Apart from the changes to `action-types`, the changes here won't affect the actual build, so this only needs a code review. There are updated tests for the reducer that you can run with:

```bash
npm run test-client client/state/purchases/
```

Although the CI indicates that the tests have passed already.

- [ ] Code

Test live: https://calypso.live/?branch=update/purchases-reducer

scruffian commented on client/state/purchases/index.js:
```diff
@@ -0,0 +1,173 @@
+ * Internal Dependencies
+ */
+import {
+	PURCHASES_REMOVE,
```
I guess we should probably alphabetize these.

scruffian replied:
Actually, why did you change this from just `
import { action as ActionTypes } from 'lib/upgrades/constants';`?

gziolo replied:
This is a common pattern used inside `state` folder.

scruffian commented on client/state/purchases/index.js:
```diff
@@ -0,0 +1,173 @@
+
+	switch ( type ) {
+		case PURCHASES_REMOVE:
+			return assign( {}, state, {
```
@gziolo says we can use spread operators in Calypso now :)

drewblaisdell replied:
Awesome.

scruffian commented on client/state/purchases/index.js:
```diff
@@ -0,0 +1,173 @@
+	return purchases;
+}
+
+export const items = ( state = initialState, action ) => {
```
Why `items`?

gziolo replied:
I think it should be named `purchases`.

drewblaisdell replied:
I used `items` because it seems to be a relatively common pattern in the state tree, e.g. [state.posts.items](https://github.com/Automattic/wp-calypso/blob/master/client/state/sites/reducer.js#L101) and [state.sites.items](https://github.com/Automattic/wp-calypso/blob/master/client/state/sites/reducer.js#L101).

scruffian commented on client/state/purchases/test/index.js:
```diff
@@ -0,0 +1,202 @@
+		siteId = 2701;
+
+	it( 'should return an object with the initial state', () => {
+		expect( items( undefined, { type: 'UNRELATED' } ) ).to.be.eql( {
```
Would it be a bad idea to import `initialState` from the reducer?

drewblaisdell replied:
I didn't consider doing this because I just copied the tests. @Tug has mentioned that he thinks this is a bad idea but I can't remember why.

scruffian commented:
This looks great. I had a few minor questions...

drewblaisdell commented:
I'm going to update this to use `createReducer` as well.

gziolo commented on client/state/action-types.js:
```diff
@@ -14,33 +14,35 @@ export const ANALYTICS_EVENT_RECORD = 'ANALYTICS_EVENT_RECORD';
 export const ANALYTICS_MULTI_TRACK = 'ANALYTICS_MULTI_TRACK';
```
Did you order those constants manually or do you use anything fancy to do it for you? :)

stephanethomas replied:
You can use <kbd>F9</kbd/> in Sublime Text to sort lines for you.

gziolo replied:
What about PhpStorm or WebStorm? :D

drewblaisdell replied:
I used `:sort` in vim.

gziolo commented on client/state/purchases/index.js:
```diff
@@ -0,0 +1,173 @@
+/**
```
This file should be renamed to `reducer.js` to follow other folders.

gziolo commented on client/state/purchases/selectors.js:
```diff
@@ -0,0 +1,7 @@
+import assign from 'lodash/assign';
+import find from 'lodash/find';
+
+export const getByPurchaseId = ( state, id ) => (
```
Wouldn't it better to name it `getByPurchaseById`?

drewblaisdell replied:
Well, there is only one criteria, so I would expect `by` to appear only once, but I might be misunderstanding the ambiguity of `getByPurchaseId`. The reason I wouldn't use `getPurchaseById` is that you can get purchase by the user or site ID as well.

gziolo replied:
I missed that first by when checking on mobile, sorry about that :)

gziolo commented:
> I'm going to update this to use createReducer as well.

It probably can wait. Let's add unit test for a selector and merge it as it as, because it's already quite big PR. Next step would be to extract actions and move action creators. Then we should be good to connect it with reducer tree and replace Flux stores.

drewblaisdell commented:
> It probably can wait. Let's add unit test for a selector and merge it as it as, because it's already quite big PR.

:+1:

I moved the reducer to `reducer.js`, updated it to use the object spread operator, added a test for the selectors, and rebased.

gziolo commented:
LGTM 👍

gziolo commented:
@drewblaisdell make sure you apply this PR #6313 properly when rebasing with master ;)

scruffian commented:
@Automattic/sdev-feed

drewblaisdell commented:
Thanks @gziolo and @scruffian.
